File: Admin Frontend/Admin Frontend/src/utils/timezone.ts
/**
 * Timezone utilities for centralized IST (Indian Standard Time) handling
 * All datetime operations should use IST timezone consistently across the application
 */

// IST timezone offset: UTC+5:30 (in minutes)
const IST_OFFSET_MINUTES = 5 * 60 + 30; // 330 minutes

/**
 * Get current datetime in IST timezone
 * @returns Date object representing current time in IST
 */
export function getISTNow(): Date {
  const now = new Date();
  // Convert UTC to IST by adding offset
  const istTime = new Date(now.getTime() + IST_OFFSET_MINUTES * 60 * 1000);
  return istTime;
}

/**
 * Convert a date string (ISO format) to IST Date object
 * @param dateString - ISO format date string (e.g., "2024-01-15T00:00:00" or "2024-01-15T00:00:00Z")
 * @returns Date object in IST
 */
export function parseToIST(dateString: string): Date {
  if (!dateString) {
    throw new Error('Invalid date string');
  }

  // Parse the date string
  let date: Date;
  if (dateString.includes('Z') || dateString.includes('+') || dateString.includes('-', 10)) {
    // ISO format with timezone
    date = new Date(dateString);
  } else {
    // Assume it's in IST if no timezone specified
    // Parse as local time and adjust to IST
    date = new Date(dateString);
  }

  // If the date string doesn't have timezone info, treat it as IST
  if (!dateString.includes('Z') && !dateString.includes('+') && !dateString.includes('-', 10)) {
    // Adjust to IST by subtracting local offset and adding IST offset
    const localOffset = date.getTimezoneOffset() * 60 * 1000;
    const istOffset = IST_OFFSET_MINUTES * 60 * 1000;
    date = new Date(date.getTime() - localOffset + istOffset);
  } else {
    // Date has timezone info, convert to IST
    const utcTime = date.getTime();
    const istTime = utcTime + IST_OFFSET_MINUTES * 60 * 1000;
    date = new Date(istTime);
  }

  return date;
}

/**
 * Format date in IST to ISO string
 * @param date - Date object (will be treated as IST)
 * @returns ISO format string
 */
export function formatISTISO(date: Date): string {
  if (!date) {
    return '';
  }

  // Convert to IST and format
  const istTime = new Date(date.getTime() + IST_OFFSET_MINUTES * 60 * 1000);
  return istTime.toISOString();
}

/**
 * Format date in IST for display
 * @param date - Date object (will be treated as IST)
 * @param options - Intl.DateTimeFormatOptions
 * @returns Formatted string
 */
export function formatISTForDisplay(
  date: Date,
  options: Intl.DateTimeFormatOptions = {
    year: 'numeric',
    month: 'short',
    day: 'numeric',
    hour: '2-digit',
    minute: '2-digit',
    timeZone: 'Asia/Kolkata',
  }
): string {
  if (!date) {
    return '';
  }

  return new Intl.DateTimeFormat('en-IN', {
    ...options,
    timeZone: 'Asia/Kolkata',
  }).format(date);
}

/**
 * Get IST timezone string
 * @returns IST timezone identifier
 */
export function getISTTimezone(): string {
  return 'Asia/Kolkata';
}

/**
 * Convert UTC date to IST
 * @param utcDate - Date object in UTC
 * @returns Date object in IST
 */
export function utcToIST(utcDate: Date): Date {
  const utcTime = utcDate.getTime();
  const istTime = utcTime + IST_OFFSET_MINUTES * 60 * 1000;
  return new Date(istTime);
}

/**
 * Convert IST date to UTC
 * @param istDate - Date object in IST
 * @returns Date object in UTC
 */
export function istToUTC(istDate: Date): Date {
  const istTime = istDate.getTime();
  const utcTime = istTime - IST_OFFSET_MINUTES * 60 * 1000;
  return new Date(utcTime);
}

/**
 * Format date as YYYY-MM-DD in IST timezone (for date comparisons)
 * This ensures dates are compared correctly without timezone shifts
 * @param date - Date object (will be treated as IST)
 * @returns Date string in YYYY-MM-DD format
 */
export function formatISTDateString(date: Date): string {
  if (!date) {
    return '';
  }

  // Format date in IST timezone to avoid timezone shifts
  const formatter = new Intl.DateTimeFormat('en-CA', {
    timeZone: 'Asia/Kolkata',
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
  });
  
  return formatter.format(date);
}

File: TGTS_Flutter (2)/TGTS_Flutter_v2/lib/utils/timezone.dart
/// Timezone utilities for centralized IST (Indian Standard Time) handling
/// All datetime operations should use IST timezone consistently across the application

import 'package:intl/intl.dart';
import 'package:timezone/timezone.dart' as tz;
import 'package:timezone/data/latest.dart' as tz;

/// IST timezone location
late tz.Location istLocation;

/// Initialize timezone data (call this once at app startup)
void initializeTimezone() {
  tz.initializeTimeZones();
  istLocation = tz.getLocation('Asia/Kolkata');
}

/// Get current datetime in IST timezone
DateTime getISTNow() {
  return tz.TZDateTime.now(istLocation);
}

/// Convert UTC DateTime to IST
DateTime utcToIST(DateTime utcDateTime) {
  final utcTZ = tz.TZDateTime.from(utcDateTime, tz.UTC);
  return utcTZ.toLocal();
}

/// Convert IST DateTime to UTC
DateTime istToUTC(DateTime istDateTime) {
  final istTZ = tz.TZDateTime.from(istDateTime, istLocation);
  return istTZ.toUtc();
}

/// Parse ISO date string and convert to IST DateTime
/// If the string doesn't have timezone info, it's assumed to be in IST
DateTime parseToIST(String dateString) {
  if (dateString.isEmpty) {
    throw ArgumentError('Invalid date string');
  }

  DateTime dateTime;
  
  // Try parsing as ISO format
  try {
    dateTime = DateTime.parse(dateString);
  } catch (e) {
    throw FormatException('Invalid date format: $dateString');
  }

  // If the date string has timezone info (Z or +/-), convert to IST
  if (dateString.contains('Z') || dateString.contains('+') || 
      (dateString.contains('-') && dateString.split('-').length > 3)) {
    // Has timezone info, convert UTC to IST
    final utcTZ = tz.TZDateTime.from(dateTime, tz.UTC);
    return tz.TZDateTime.from(utcTZ, istLocation);
  } else {
    // No timezone info, assume it's already in IST
    // Create TZDateTime in IST timezone
    return tz.TZDateTime(
      istLocation,
      dateTime.year,
      dateTime.month,
      dateTime.day,
      dateTime.hour,
      dateTime.minute,
      dateTime.second,
      dateTime.millisecond,
      dateTime.microsecond,
    );
  }
}

/// Format DateTime in IST to ISO string
String formatISTISO(DateTime dateTime) {
  tz.TZDateTime istTZ;
  if (dateTime is tz.TZDateTime) {
    istTZ = tz.TZDateTime.from(dateTime, istLocation);
  } else {
    istTZ = tz.TZDateTime(
      istLocation,
      dateTime.year,
      dateTime.month,
      dateTime.day,
      dateTime.hour,
      dateTime.minute,
      dateTime.second,
      dateTime.millisecond,
      dateTime.microsecond,
    );
  }
  return istTZ.toIso8601String();
}

/// Format DateTime in IST for display
/// [locale] - locale string (e.g., 'en', 'te')
/// [format] - date format string (e.g., 'dd MMM yyyy, hh:mm a')
String formatISTForDisplay(DateTime dateTime, {String? locale, String? format}) {
  tz.TZDateTime istTZ;
  if (dateTime is tz.TZDateTime) {
    istTZ = tz.TZDateTime.from(dateTime, istLocation);
  } else {
    istTZ = tz.TZDateTime(
      istLocation,
      dateTime.year,
      dateTime.month,
      dateTime.day,
      dateTime.hour,
      dateTime.minute,
      dateTime.second,
      dateTime.millisecond,
      dateTime.microsecond,
    );
  }
  
  if (format != null) {
    final formatter = DateFormat(format, locale ?? 'en');
    return formatter.format(istTZ);
  }
  
  // Default format
  final formatter = DateFormat('dd MMM yyyy, hh:mm a', locale ?? 'en');
  return formatter.format(istTZ);
}

/// Get IST timezone identifier
String getISTTimezone() {
  return 'Asia/Kolkata';
}

/// Ensure DateTime is in IST timezone
/// If the DateTime is naive (no timezone), it's assumed to be in IST
tz.TZDateTime ensureIST(DateTime dateTime) {
  if (dateTime is tz.TZDateTime) {
    return dateTime.toLocal();
  }
  return tz.TZDateTime.from(dateTime, istLocation);
}

